### PR TITLE
[quant] Implemented InputWeightObserver for Linear inputs

### PR DESCRIPTION
**Pull request by angelayi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59474 [quant] Input Weight Equalization - prepare modifications
* **#59090 [quant] Implemented InputWeightObserver for Linear inputs**

Summary: Implemented two observers (InputEqualObserver and WeightEqualObserver) which will be inserted into the graph during prepare_fx().

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

### CODE REVIEW

File: torch/quantization/observer.py [hunk 1]
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
         return x_orig
 
 
+class InputWeightMinMaxObserver(PerChannelMinMaxObserver):
```
Comment (supriyar) [inline torch/quantization/observer.py:805]: Let us prefix this with `_` since we don't want to expose this to the user. It might also be a good idea to put this into a separate file with the rest of the equalization stuff.

File: torch/quantization/observer.py [hunk 2]
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
...
+    max_inputs_col: torch.Tensor
+    min_weights_col: torch.Tensor
+    max_weights_col: torch.Tensor
+    min_weights_row: torch.Tensor
+    max_weights_row: torch.Tensor
+
+    def __init__(self, dtype=torch.quint8,
+                 qscheme=torch.per_channel_affine, reduce_range=False,
+                 quant_min=None, quant_max=None, factory_kwargs=None) -> None:
+        super(InputWeightMinMaxObserver, self).__init__(ch_axis=1, dtype=dtype,
```
Comment (supriyar) [inline torch/quantization/observer.py:851]: Are we using the ch_axis field anywhere in the observer?

File: torch/quantization/observer.py [hunk 3]
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
...
+                                                        qscheme=qscheme,
```
Comment (supriyar) [inline torch/quantization/observer.py:852]: I think qscheme is confusing here because we don't know if it corresponds to input or weight. Can we distinguish it? In the code we can assert that we only expect per-tensor for input.

File: torch/quantization/observer.py (hunk 2 above)
Comment (supriyar) [reply torch/quantization/observer.py:851]: let us also have an input_dtype and weight_dtype to separate them out

File: torch/quantization/observer.py [hunk 4]
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
...
+        self.ch_axis = 0
+        min_weights_row, max_weights_row = self._find_minmax(w_orig, self.min_weights_row, self.max_weights_row)
+        self.min_weights_row.resize_(min_weights_row.shape)
+        self.max_weights_row.resize_(max_weights_row.shape)
+        self.min_weights_row.copy_(min_weights_row)
+        self.max_weights_row.copy_(max_weights_row)
+
+        return x_orig, w_orig
+
+    def calculate_scale(self) -> torch.Tensor:
```
Comment (supriyar) [inline torch/quantization/observer.py:908]: let's call this equalization_scale instead to not confuse with `scale` from qparams

File: torch/quantization/observer.py [hunk 5]
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
...
+        Returns the scale/zero_point for the input and weight rows
+        """
+
+        # Calculate qparams for the scaled inputs
+        min_input_scaled = torch.mul(torch.min(self.min_inputs_col), self.S)
+        max_input_scaled = torch.mul(torch.max(self.max_inputs_col), self.S)
+        (scale_input, zero_point_input) = self._calculate_qparams(min_input_scaled, max_input_scaled)
+
+        # Calculate qparams for the weights
+        (scale_weight, zero_point_weight) = self._calculate_qparams(self.min_weights_row, self.max_weights_row)
```
Comment (supriyar) [inline torch/quantization/observer.py:923]: we'd also need to multiply the inverse scale factors to the weights.

File: torch/quantization/observer.py [hunk 6]
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
...
+
+        # Calculate the min/max of weight columns
+        min_weights_col, max_weights_col = self._find_minmax(w_orig, self.min_weights_col, self.max_weights_col)
+        self.min_weights_col.resize_(min_weights_col.shape)
+        self.max_weights_col.resize_(max_weights_col.shape)
+        self.min_weights_col.copy_(min_weights_col)
+        self.max_weights_col.copy_(max_weights_col)
+
+        # Calculate the min/max of weight rows
+        self.ch_axis = 0
```
Comment (supriyar) [inline torch/quantization/observer.py:899]: rather than doing this, it might be a better idea to take an axis argument in _find_minmax

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

File: torch/quantization/fx/_equalize.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        - input_obs: Used to record the running minimum and maximum of the
+        columns of incoming input tensors
+        - weight_col_obs: Used to record the running minimum and maximum of
+        columns of incoming weight tensors
+        - weight_row_obs: Used to record the running minimum and maximum of
+        rows of incoming weight tensors
+
+    Given running min/max of the input columns as :math:`x_\text{min}` and :math:`x_\text{max}`,
+    and the running min/max of the weight columns as :math:`w_\text{min}` and :math:`w_\text{max}`,
+    scale :math:`S` is computed as:
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:40]: nit: rename scale -> equalization_scale

File: torch/quantization/fx/_equalize.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    The running minimum/maximum :math:`x_\text{min/max}` and
+    :math:`w_\text{min/max}` are computed in the same way as
+    :class:`~torch.quantization.observer.MinMaxObserver`, with the difference
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:44]: Should this be PerChannelMinMaxObserver?

File: torch/quantization/fx/_equalize.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    that the running min/max values are stored per column.
+
+    The scale factor :math:`S` is then computed as:
+
+    .. math::
+        S = \sqrt{\frac{x_{max} - x_{min}}{w_{max} - w{min}}}
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:50]: Shouldn't the math be `\sqrt{\frac{w_{max} - w_{min}}{x_{max} - x{min}}}`

File: torch/quantization/fx/_equalize.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+    where :math:`X` is the observed tensor.
+
+    .. note:: If the running minimum equals to the running maximum, the scales
+              and zero_points are set to 1.0 and 0.
+    """
+
+    def __init__(self, input_dtype=torch.quint8, input_qscheme=torch.per_tensor_affine,
+                 input_quant_min=None, input_quant_max=None, weight_dtype=torch.quint8,
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:59]: I think weight dtype should be torch.qint8

File: torch/quantization/fx/_equalize.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        x_orig = self.input_obs(x_orig)
+        w_orig = self.weight_col_obs(w_orig)
+        w_orig = self.weight_row_obs(w_orig)
+
+        # Calculate the column indices of the min/max weight in each row
+        num_row, _ = w_orig.shape
+        min_weights_ind = []
+        max_weights_ind = []
+        for i in range(num_row):
+            min_weights_ind.append(torch.where(w_orig[i] == self.weight_row_obs.min_vals[i])[0][0])
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:111]: torch.where(cond) is equivalent to torch.nonzero(condition)
https://pytorch.org/docs/stable/generated/torch.nonzero.html#torch.nonzero

File: torch/quantization/fx/_equalize.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+            warnings.warn(
+                "Must find the column indicies of the minimum of each row in the \
+                weights in order to calculate the qparams calculate the \
+                qparams. Returning default scale and zero point. "
+            )
+            return torch.tensor([1.0]), torch.tensor([0]), torch.tensor([1.0]), torch.tensor([0])
+
+        (min_inputs, max_inputs) = self.get_input_minmax()
+        min_input_ind = torch.argmin(min_inputs)
+        max_input_ind = torch.argmax(max_inputs)
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:156]: let's multiply the equalization_scale with the min_inputs/max_inputs and then compute the new min/max for the inputs.
So we can do torch.argmin(min_inputs * eq_scale) instead of torch.argmin(min_inputs) * eq_scale[idx]

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

File: torch/quantization/fx/_equalize.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        x_orig = self.input_obs(x_orig)
+        w_orig = self.weight_col_obs(w_orig)
+        w_orig = self.weight_row_obs(w_orig)
+
+        # Calculate the column indices of the min/max weight in each row
+        num_row, _ = w_orig.shape
+        min_weights_ind = []
+        max_weights_ind = []
+        for i in range(num_row):
+            min_weights_ind.append(torch.nonzero(w_orig[i] == self.weight_row_obs.min_vals[i])[0][0])
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:111]: Do we still need to access [0][0] even after using `torch.nonzero`?

Comment (angelayi) [reply torch/quantization/fx/_equalize.py:111]: Yes, torch.nonzero seems to return an element in a 2D array.

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

File: test/quantization/fx/test_equalize_fx.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+                                           torch.per_channel_affine_float_qparams)))
+    def test_input_weight_observer(self, input_qdtype, input_qscheme, weight_qdtype, weight_qscheme):
+        input_obs = _InputEqualObserver(dtype=input_qdtype,
+                                        qscheme=input_qscheme)
+        weight_obs = _WeightEqualObserver(dtype=weight_qdtype,
+                                          qscheme=weight_qscheme)
+
+        width = np.random.randint(10) + 1
+        x_height = np.random.randint(10) + 2
+        w_height = np.random.randint(10) + 2
```
Comment (supriyar) [inline test/quantization/fx/test_equalize_fx.py:29]: Why do we add the constant values here?

Comment (angelayi) [reply test/quantization/fx/test_equalize_fx.py:29]: The constant values are to make sure that the width is greater than 0 and the heights are greater than 1. If the heights are equal to 1 then the min/max values will be the same and there will be a division by 0 error when calculating the scale.

Comment (angelayi) [reply test/quantization/fx/test_equalize_fx.py:29]: I can also replace them with `w = np.random.randint(1, 10)` and `h = np.random.randint(2, 10)` if that's more readable.

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

File: test/quantization/fx/test_equalize_fx.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
+import torch
+from torch.testing._internal.common_quantization import QuantizationTestCase
+from torch.quantization.fx._equalize import (
+    _InputEqualObserver, _WeightEqualObserver, calculate_equalization_scale
```
Comment (jerryzh168) [inline test/quantization/fx/test_equalize_fx.py:4]: nit: InputEqualizationObserver and WeightEqualizationObserver feels clearer..

File: torch/quantization/fx/_equalize.py [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+        # located at the same column index
+        (min_weights, max_weights) = self.get_weight_row_minmax()
+        min_weights_scaled = torch.mul(min_weights, torch.reciprocal(self.equalization_scale[self.min_weights_ind]))
+        max_weights_scaled = torch.mul(max_weights, torch.reciprocal(self.equalization_scale[self.max_weights_ind]))
+        (scale_weight, zero_point_weight) = self.weight_row_obs._calculate_qparams(min_weights_scaled, max_weights_scaled)
+
+        return scale_weight, zero_point_weight
+
+
+def calculate_equalization_scale(input_obs: _InputEqualObserver, weight_obs: _WeightEqualObserver) -> torch.Tensor:
```
Comment (jerryzh168) [inline torch/quantization/fx/_equalize.py:198]: should we set_equalization_scale in convert instead? setting equalization scale in a `calculate_equalization_scale` doesn't sound very intuitive.

File: torch/quantization/fx/_equalize.py [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+    """
+
+    def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
+                 quant_min=None, quant_max=None, factory_kwargs=None) -> None:
+        super(_InputEqualObserver, self).__init__()
+
+        if qscheme not in {torch.per_tensor_affine, torch.per_tensor_symmetric}:
+            raise TypeError("Input qscheme must be per-tensor")
+
+        self.input_obs = PerChannelMinMaxObserver(ch_axis=1, dtype=dtype,
```
Comment (jerryzh168) [inline torch/quantization/fx/_equalize.py:40]: I think we should have an output_obs as well for output observer, right?

File: torch/quantization/fx/_equalize.py (hunk 16 above)
Comment (angelayi) [reply torch/quantization/fx/_equalize.py:198]: Sure, I will move it out and instead set the equalization scale in convert().

File: torch/quantization/fx/_equalize.py (hunk 17 above)
Comment (angelayi) [reply torch/quantization/fx/_equalize.py:40]: I thought the output observer was taken into account in the from_float() function and we would just replace the InputEqualizationObserver with a MinMaxObserver in prepare()?

Comment (supriyar) [reply torch/quantization/fx/_equalize.py:40]: The output observer is set to whatever is used for the input observer currently. So if the user wants to use something other than MinMax for the output, they wouldn't be able to.
We can add an optional argument to the InputEqualizationObserver that takes in output_observer and set it to default MinMaxObserver.

Comment (angelayi) [reply torch/quantization/fx/_equalize.py:40]: Oh ok, got it!

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

**Review by supriyar — APPROVED** [review]

File: torch/quantization/fx/_equalize.py [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+                                             factory_kwargs=factory_kwargs)
+        else:
+            self.output_obs = output_obs
+
+        self.equalization_scale = torch.empty(0)
+
+    def forward(self, x_orig):
+        # TODO: Allow for convoluational layers
+        if not (x_orig.ndim == 2):
+            raise ValueError("InputEqualObserver only supports Linear layers")
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:63]: Update the message here to InputEqualizationObserver

File: torch/quantization/fx/_equalize.py [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+                                                       quant_min=quant_min,
+                                                       quant_max=quant_max,
+                                                       factory_kwargs=factory_kwargs)
+
+        self.equalization_scale = torch.empty(0)
+
+    def forward(self, w_orig):
+        # TODO: Allow for convoluational layers
+        if not (w_orig.ndim == 2):
+            raise ValueError("WeightEqualObserver only supports Linear layers")
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:146]: same here too

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

**Review by jerryzh168 — APPROVED** [review]
Looks good, thanks for addressing my comments!

File: torch/quantization/fx/_equalize.py [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+        if qscheme not in {torch.per_tensor_affine, torch.per_tensor_symmetric}:
+            raise TypeError("Input qscheme must be per-tensor")
+
+        self.input_obs = PerChannelMinMaxObserver(ch_axis=1, dtype=dtype,
+                                                  qscheme=qscheme,
+                                                  quant_min=quant_min,
+                                                  quant_max=quant_max,
+                                                  factory_kwargs=factory_kwargs)
+
+        if output_obs is None:
```
Comment (jerryzh168) [inline torch/quantization/fx/_equalize.py:49]: actually I think this should probably be a constructor as well? otherwise all the Tensors in the model will use the same output observer if user customizes it.

**Comment by bigfootjon** [comment]
Closed by https://github.com/pytorch/pytorch/commit/cc03ea2c476b1137fb26f7fffac3b62dfcb6329c
